### DISCUSSION TITLE
Idea: Preserve matchdata in rules

### DISCUSSION THREAD

**Discussion by avdi** [ruby/rake#437] [Ideas]
Here's part of a rule I just wrote the other day:

``` ruby
stampfile_pattern = %r"^members/(?<email>.*)/\d{2}-(?<step>\w+).timestamp$"
rule stampfile_pattern do |t|
  match           = stampfile_pattern.match(t.name)
  email           = match[:email]
  step            = match[:step]
  # ...
end
```

Notice how, after it matches, I have to re-apply the regex inside the rule definition in order to pull out named matchdata.

I was thinking of introducing something like this as an alternative:

``` ruby
stampfile_pattern = %r"^members/(?<email>.*)/\d{2}-(?<step>\w+).timestamp$"
rule stampfile_pattern do |t|
  email           = t.matchdata[:email]
  step            = t.matchdata[:step]
  # ...
end
```

Before I dive in, what do you think? And is there another way to accomplish this that I've missed?

**Comment by drbrain** [comment]
I think this would be a nice addition.

I don't think there's a way to accomplish this without adding the feature to the rule.

I've not used a rule without dependencies, it gets triggered by a separate task that depends on the output (regexp)?

**Comment by avdi** [comment]
> I don't think there's a way to accomplish this without adding the feature to the rule.

Yeah, that's what I was thinking. I feel like it makes sense, since the task object already has the concrete `t.name` that matched, to also have the matchdata.

> I've not used a rule without dependencies, it gets triggered by a separate task that depends on the output (regexp)?

Yep.

**Comment by avdi** [comment]
I've also thought about this:

``` ruby
rule %r"^members/(?<email>.*)/\d{2}-(?<step>\w+).timestamp$" => 
  "members/%{email}/some-input-file" do |t|
  email           = t.matchdata[:email]
  step            = t.matchdata[:step]
  # ...
end
```

I.e. adding the ability to substitute in backreferences in the rule source. Currently, rule sources suffer from the same problem as the rule body: if you want to reference regex matchdata, you have to use a lambda as the dependency and re-apply the regex.

**Comment by yeban** [comment]
I stumbled upon this issue looking for how to access `matchdata` inside a rule. This would be a great addition, imho.

**Comment by avdi** [comment]
Oh hey, thanks for reminding me.
